posthog/temporal/utils.py

```diff
@@ -0,0 +1,35 @@
+import asyncio
+from functools import wraps
+from typing import Any, Awaitable, Callable, TypeVar, cast
+
+from temporalio import activity
+
+F = TypeVar("F", bound=Callable[..., Awaitable[Any]])
+
+
+def auto_heartbeater(fn: F) -> F:
+    # We want to ensure that the type hints from the original callable are
+    # available via our wrapper, so we use the functools wraps decorator
+    @wraps(fn)
+    async def wrapper(*args, **kwargs):
+        heartbeat_timeout = activity.info().heartbeat_timeout
+        heartbeat_task = None
+        if heartbeat_timeout:
+            # Heartbeat twice as often as the timeout
+            heartbeat_task = asyncio.create_task(heartbeat_every(heartbeat_timeout.total_seconds() / 2))
+        try:
+            return await fn(*args, **kwargs)
+        finally:
+            if heartbeat_task:
+                heartbeat_task.cancel()
+                # Wait for heartbeat cancellation to complete
+                await asyncio.wait([heartbeat_task])
+
+    return cast(F, wrapper)
```

Comment:
A bit nitpicky but `cast` doesn't do anything: 

> This returns the value unchanged. To the type checker this signals that the return value has the designated type, but at runtime we intentionally don’t check anything

I'd prefer if we used the appropriate type hints instead as `cast` could be hiding something.